rmw_zenoh_cpp/src/detail/rmw_service_data.cpp

```diff
@@ -344,12 +344,12 @@ rmw_ret_t ServiceData::take_request(
   request_header->received_timestamp = now_ns.count();
 
   // Add this query to the map, so that rmw_send_response can quickly look it up later.
-  const size_t hash = rmw_zenoh_cpp::hash_gid(request_header->request_id.writer_guid);
-  std::unordered_map<size_t, SequenceToQuery>::iterator it = sequence_to_query_map_.find(hash);
+  std::unordered_map<size_t, SequenceToQuery>::iterator it =
+    sequence_to_query_map_.find(entity_->keyexpr_hash());
```

Comment:
I'm confused by this change and the related ones below.

The `sequence_to_query_map_` maps the client's GID hash to a map of sequence number from the same client to the `ZenohQuery`. Line 355 highlights highlights this. Also see [this line](https://github.com/ros2/rmw_zenoh/blob/1fb2b1e903c29c774525c84d9c9d5be8badb3cc8/rmw_zenoh_cpp/src/rmw_zenoh.cpp#L1787-L1788) from `rmw_send_request` that appends the client's GID to the header.

But here we're always using the service entity's GID hash to lookup `sequence_to_query_map_`. imo we still need to call `rmw_zenoh_cpp::hash_gid(request_header->request_id.writer_guid)`. See my comment above on retaining this function.

Reply:
You are right; commit eb2239213168d6ebcc897363f7fcf31c77a186bf is completely wrong (I wonder how it worked at all in my testing).  I've now reverted it in 728c072d9b488d274b95c3843b09b884a95e877a .